File: BackUpForABook/UIControls/NGClasses/ANNodeLayer.m
//
//  ASymbol_Brain.m
//  AnnieVisualyticsPageView
//
//  Created by hui wang on 9/11/15.
//  Copyright (c) 2015 AnnieLyticx. All rights reserved.
//

#import "ANNodeLayer.h"

@implementation ANNodeLayer
{
    @private

    CAShapeLayer *_imageLayer;
}

-(void)initWithDefault
{
    _imageLayer = [CAShapeLayer layer];
    _imageLayer.contents = (id) [UIImage imageNamed:@"face_red@3x.png"].CGImage;
    _imageLayer.masksToBounds = YES;
    _imageLayer.borderColor = [UIColor clearColor].CGColor;
    _imageLayer.backgroundColor = [UIColor clearColor].CGColor;
}

// !!! The view area is CGRectMake(0, 0, self.bounds.size.width, self.bounds.size.height)

- (void)drawInContext:(CGContextRef)ctx
{
    CGContextSaveGState(ctx);
    UIGraphicsPushContext(ctx);
    
    UIBezierPath* rectanglePath = [UIBezierPath bezierPathWithRect: CGRectMake(0, 0, self.bounds.size.width, self.bounds.size.height)];
    CGContextSetLineWidth(ctx, 2.0);
    
    [rectanglePath fill];
    
    UIGraphicsPopContext();
    CGContextRestoreGState(ctx);
}

-(void)setFrame:(CGRect)frame
{
    _imageLayer.frame = frame;
}

-(void)loadOntoPage:(CALayer*)parentPage
             layout:(CGRect)layout
{
    // map to _imageLayer
    if (_imageLayer && _imageLayer.superlayer)
    {
        [_imageLayer removeFromSuperlayer];
    }
    
    // connect to the new parent
    [parentPage addSublayer:_imageLayer];
    
    // set frame
    _imageLayer.frame = layout;
}


@end


ANNodeLayer* getRandomSymbolPath()
{
    ANNodeLayer* obj = [[ANNodeLayer alloc] init];
    
    [obj initWithDefault];
    
    return obj;
}

ANNodeLayer* getRandomSymbolPath_Face()
{
    ANNodeLayer* obj = [[ANNodeLayer alloc] init];
    
    [obj initWithDefault];
    
    return obj;
}

File: iOS/demos/NGClasses_swift/ANNodeLayer.swift
//
//  ANNodeLayer.swift
//  NGClasses
//
//  Created by iMac on 16/12/20.
//  Copyright © 2016年 LyTsai. All rights reserved.
//

import Foundation
import UIKit

// MARK: ------------ NANodeLayer ------------------
class ANNodeLayer: CALayer {
    // this layer is not added to any superLayer
    // the "isHidden" property of CALayer, used with ANLayoutNode's isHidden
    
    var imageLayer = CALayer()
    var textLayer = CATextLayer()
    
    // zero, no changes; (0.5, 0.5), halfWidth and halfHeight
    var shrinkRatio: UIOffset = UIOffset.zero
    // adjust of imageLayer and textLayer
    // TODO: ------ in case of more detailed change, leave it now
    var shrink: CGFloat = 0
    
    class func createWithImage(_ image: CGImage!, text: String!) -> ANNodeLayer {
        let nodeLayer = ANNodeLayer()
        nodeLayer.imageLayer.contents = image ?? UIImage(named: "flowers")!.cgImage
        nodeLayer.textLayer.string = text
        nodeLayer.textLayer.foregroundColor = UIColor.black.cgColor
        nodeLayer.textLayer.alignmentMode = "center"
        
        return nodeLayer
    }
}

// layout and add
extension ANNodeLayer {
    // load into parent page and set up transformation matrix that would map the original view onto given layout
    // parentPage -  our new parent layer
    // layout - the new frame area
    func loadOntoPage(_ parentPage: CALayer, layout: CGRect) {
        // map to _imageLayer
        if  imageLayer.superlayer != nil {
            imageLayer.removeFromSuperlayer()
        }
        // connect to the new parent
        parentPage.addSublayer(imageLayer)
        
        if  textLayer.superlayer != nil {
            textLayer.removeFromSuperlayer()
        }
        // connect to the new parent
        parentPage.addSublayer(textLayer)
        
        // set frame
        bounds = layout.insetBy(dx: layout.width * 0.5 * shrinkRatio.horizontal, dy: layout.height * 0.5 * shrinkRatio.vertical)
        
        layoutSublayers()
    }
    
    override func layoutSublayers() {
        super.layoutSublayers()
        
        // the relationship between imageLayer-bounds, textLayer-bounds
        imageLayer.frame = bounds.insetBy(dx: bounds.width * shrink * 0.5, dy: bounds.height * shrink * 0.5)
        textLayer.frame = bounds.insetBy(dx: bounds.width * shrink * 0.5, dy: bounds.height * shrink * 0.5)
//        textLayer.font = UIFont.systemFont(ofSize: 10)
    }
}

// MARK: ------------- default for test ---------------------
extension ANNodeLayer {
    func initWithDefault()  {
        imageLayer.contents = UIImage(named: "face_red")!.cgImage
        imageLayer.masksToBounds = true
        imageLayer.borderColor = UIColor.clear.cgColor
        imageLayer.backgroundColor = UIColor.clear.cgColor
    }
    func initWithDefault1()  {
        imageLayer.contents = UIImage(named: "face1")!.cgImage
        imageLayer.masksToBounds = true
        imageLayer.borderColor = UIColor.clear.cgColor
        imageLayer.backgroundColor = UIColor.clear.cgColor
    }
}

func getRandomSymbolPath() -> ANNodeLayer {
    let obj = ANNodeLayer()
    obj.initWithDefault()
    
    return obj
}

func getRandomSymbolPath_Face() -> ANNodeLayer {
    let obj = ANNodeLayer()
    obj.initWithDefault1()
    
    return obj
}
